### PR TITLE
feat: multiple file support for file_drop

### PR DESCRIPTION
**Pull request by hkayabilisim**
Added multiple file support to file_drop component.  
Same as the original component, directories are simply ignored. Multiple file support was actually embedded in the underlying VUE scripts. I've done minor changes. 

**Compatibility**:
on_file callback function now accepts List[FileInfo] as opposed to single FileInfo object. This may bring some problems for existing codes. As a remedy, if the number of files is just one, we can return a single FileInfo. If not, we return list. I leave it up to you to decide.

This is related to: https://github.com/widgetti/solara/issues/260

### CODE REVIEW

File: solara/components/file_drop.vue [hunk 1]
```diff
@@ -1,6 +1,13 @@
 <template>
   <div ref="dropzone" class="solara-file-drop" effectAllowed="move">
-    {{ (file_info && file_info.length === 1) ? file_info[0].name : label }}
+    <div v-if="file_info && file_info.length">
```
Comment (mariobuikhuizen) [inline solara/components/file_drop.vue:3]: Ideally we would also keep the HTML the same for the single file case, as users might have CSS rules that depend on that structure. You can use `<template v-if=`, this wil not produce an HTML element in the output.

File: solara/components/file_drop.vue [hunk 2]
```diff
@@ -1,6 +1,13 @@
...
+      <ul>
+        <li v-for="file in file_info">{{ file.name }}  {{ file.size }} bytes</li>
```
Comment (mariobuikhuizen) [inline solara/components/file_drop.vue:5]: I would like the list to not have bullet points.

**Review by mariobuikhuizen — CHANGES_REQUESTED** [review]
Thanks for your contribution @hkayabilisim!

We would like to keep this component backward compatible. We could accomplish this by adding an argument `multiple` with a default value of `False` and have on_file also work with a single file (see [this](https://github.com/widgetti/solara/blob/e3ea88aa617b30da89d8fe92682c7346065d70e4/solara/components/input.py#L141) for an example of the typing, note the argument `optional: Literal[False]`)

File: solara/components/file_drop.vue [hunk 3]
```diff
@@ -16,16 +23,17 @@ module.exports = {
...
+      this.native_file_info = nativeFiles
       this.file_info = this.native_file_info.map(
-          ({name, isFile, size}) => ({
-            name,
-            isFile,
-            size,
-          }));
+        ({name, isFile, size}) => ({
+          name,
+          isFile: !!isFile, // Ensure this is a boolean
```
Comment (mariobuikhuizen) [inline solara/components/file_drop.vue:34]: Why is this needed? The spec says it's a boolean: https://developer.mozilla.org/en-US/docs/Web/API/FileSystemEntry/isFile

File: solara/components/file_drop.vue (hunk 1 above)
Comment (hkayabilisim) [reply solara/components/file_drop.vue:3]: I guess, this one will solve the problem.

~~~vue
<template>
  <div ref="dropzone" class="solara-file-drop" effectAllowed="move">
    <template v-if="file_info && file_info.length === 1"> 
      {{ file_info[0].name }}
    </template>
    <template v-else-if="file_info && file_info.length > 1"> 
      <div v-for="file in file_info">
        {{ file.name }}
      </div>
    </template>
    <template v-else> 
      {{ label }}
    </template>
  </div>
</template>
~~~

File: solara/components/file_drop.vue (hunk 2 above)
Comment (hkayabilisim) [reply solara/components/file_drop.vue:5]: Done!

File: solara/components/file_drop.vue (hunk 3 above)
Comment (hkayabilisim) [reply solara/components/file_drop.vue:34]: Yes `FileSystemEntry` has `isFile` property. However,  after applying Promises (which effectively creates File objects), we don't have isFile property anymore. So isFile of the returned object was undefined. This was true for the original version. By using !!isFile, I was trying to create a boolean variable from undefined. Anyway, it turns out that isFile is not important because it is never used after this place. So this bug was not causing any problem. 

I'm suggesting the following. I think it is ok to set isFile=true because in the beginning we already filter files.

~~~javascript
this.file_info = this.native_file_info.map(
  ({name, size}) => ({
    name,
    isFile: true,
    size,
  }));
~~~

**Review by hkayabilisim — COMMENTED** [review]
missing ';' at the end of the "padding:" line was causing CSS problem.

**Review by hkayabilisim — COMMENTED** [review]
Changes required for backward compatibility:

1. multiple property is added to FileDrop
2. Two different on_file callback signature used: FileInfo and List[FileInfo]
3. FileDrop is overloaded to match two different on_file signatures.
4. FileDrop apidoc is updated
5. FileDrop example is updated
6. bullet list in VUE template is removed
7. isFile property is always set to True (this was a hidden bug)
8. app/scatter.py is reverted to original version

**Review by hkayabilisim — COMMENTED** [review]
Comments of @mariobuikhuizen are handled in [7d44858ec65c103c299d9757c4ef169e5f5b9f9b](https://github.com/widgetti/solara/pull/562/commits/7d44858ec65c103c299d9757c4ef169e5f5b9f9b)

**Comment by maartenbreddels** [comment]
Looking good!
A few issues

- due to a large refactor, files have moved (see https://github.com/widgetti/solara/pull/421)
- I think we need to pass 'multiple' to the frontend, because now you can still drag in multiple files and they render in the component, I think we should show only 1 
- Some style issues (if you install pre-commit and run `pre-commit install`, those will be fixed for you - and we should document this on https://solara.dev/docs/howto/contribute )

Looking forward to see this green in CI, and get this merged 🥳

**Comment by hkayabilisim** [comment]
Used "multiple" option in file_drop.vue to show single file when multiple is not selected.

**Comment by hkayabilisim** [comment]
Hi @maartenbreddels 

There is only one error I couldn't resolve. Help needed!

image

**Comment by iisakkirotko** [comment]
Hi @hkayabilisim! Again, thanks for the contribution. 

The remaining issue is due to the how the component is used in the documentation: `multiple_upload` is a variable, and therefore not a Literal, so it doesn't match any of the overload variants. This can be fixed by adding an overload variant which takes a `bool` instead of `Literal[...]` for `multiple`, i.e. 

```python
@overload
@solara.component
def FileDrop(
    label=...,
    on_total_progress: Optional[Callable[[float], None]] = ...,
    on_file: Optional[Callable[[Union[FileInfo, List[FileInfo]]], None]] = ...,
    lazy: bool = ...,
    multiple: bool = ...,
) -> ipyvuetify.extra.FileInput:
    ...
```

Then the two functions that are currently used in the documentation could be combined to one that takes a Union of `FileInfo` and `List[FileInfo]` as an argument, like so:

```python
def process_files(files: Union[FileInfo, List[FileInfo]]):
    if not isinstance(files, list):
        files = [files]
    set_filename([f["name"] for f in files])
    set_size([f["size"] for f in files])
    set_content([f["file_obj"].read(100) for f in files])
```

File: solara/components/file_drop.py [hunk 4]
```diff
@@ -25,25 +26,51 @@ class FileDropZone(FileInput):
...
+@overload
+@solara.component
+def FileDrop(
+    label: str = ...,
+    on_total_progress: Optional[Callable[[float], None]] = ...,
+    on_file: Optional[Callable[[List[FileInfo]], None]] = None,
+    lazy: bool = ...,
+    multiple: Literal[True] = ...,
+) -> ipyvuetify.extra.FileInput:
+    ...
```
Comment (iisakkirotko) [inline solara/components/file_drop.py:54]: Hi @hkayabilisim! Again, thanks for the contribution. 

The remaining issue is due to the how the component is used in the documentation: `multiple_upload` is a variable, and therefore not a Literal, so it doesn't match any of the overload variants. This can be fixed by adding an overload variant which takes a `bool` instead of `Literal[...]` for `multiple`, i.e. 

```suggestion
@overload
@solara.component
def FileDrop(
    label=...,
    on_total_progress: Optional[Callable[[float], None]] = ...,
    on_file: Optional[Callable[[Union[FileInfo, List[FileInfo]]], None]] = ...,
    lazy: bool = ...,
    multiple: bool = ...,
) -> ipyvuetify.extra.FileInput:
    ...
```

File: solara/website/pages/documentation/components/input/file_drop.py [hunk 5]
```diff
@@ -10,25 +11,32 @@
...
-        set_size(file["size"])
-        f = file["file_obj"]
-        set_content(f.read(100))
+    def process_multiple_files(files: List[FileInfo]):
+        set_filename([f["name"] for f in files])
+        set_size([f["size"] for f in files])
+        set_content([f["file_obj"].read(100) for f in files])
+
+    def process_single_file(file: FileInfo):
+        process_multiple_files([file])
```
Comment (iisakkirotko) [inline solara/website/pages/documentation/components/input/file_drop.py:25]: Then the two functions that are currently used in the documentation could be combined to one that takes a Union of `FileInfo` and `List[FileInfo]` as an argument, like so:

```suggestion
def process_files(files: Union[FileInfo, List[FileInfo]]):
    if not isinstance(files, list):
        files = [files]
    set_filename([f["name"] for f in files])
    set_size([f["size"] for f in files])
    set_content([f["file_obj"].read(100) for f in files])
```

File: solara/website/pages/documentation/components/input/file_drop.py [hunk 6]
```diff
@@ -10,25 +11,32 @@
...
 
     with solara.Div() as main:
+        solara.Checkbox(label="Multiple upload", value=multiple_upload, on_value=set_multiple_upload)
         solara.FileDrop(
-            label="Drag and drop a file here to read the first 100 bytes",
-            on_file=on_file,
+            label="Drag and drop files(s) here to read the first 100 bytes.",
+            on_file=process_multiple_files if multiple_upload else process_single_file,
```
Comment (iisakkirotko) [inline solara/website/pages/documentation/components/input/file_drop.py:31]: And

```suggestion
            on_file=process_files,
```

**Comment by maartenbreddels** [comment]
Hi Huseyin,

we had quite a long discussion internally on this PR.
For ToggleButton https://github.com/widgetti/solara/blob/master/solara/components/togglebuttons.py we had two different components
 * ToggleButtonsSingle
 * ToggleButtonsMultiple

This is because otherwise the typing becomes quite complex (as you can see now for FileDrop), and it only works for literals. So if you want to take advantage of the typing you need to do:
```
if multiple:
  FileDrop(..., multiple=True)
else:
  FileDrop(..., multiple=False)
```

Which is a bit odd. For this reason, it might be better to have
`FileDrop` and `FileDropMultiple` (they can call the same internal component btw, it's just the public API)

However, also that would be inconsistent in the naming, so either solution is not perfect.

Two questions:
 1. What do you think about the typing and one vs two components
 2. If you also prefer FileDropMultiple, are you willing to do the changes? If not, we're happy to help.

Let us know, and thanks you for your help on this.

Regards,

Maarten

**Comment by hkayabilisim** [comment]
Hi Maarten,

Getting rid of some complexity by creating two variants (FileDrop and FileDropMultiple) seems a good idea. 

I've just created FileDropMultiple and commited to the branch and tested little bit (https://github.com/widgetti/solara/pull/562/commits/fa6cd2367b445d674e4d89ccc827f386d60bf7ac). It seems ok, and CI checks are green. Yet, you should revise my implementation in case if I missed some points. 

Sincerely
Huseyin

**Comment by maartenbreddels** [comment]
Hi Huseyin,

whow, many thank for doing all this work!
CI looks green (ignore that failing one, that's not related to what you do)
Will review after the weekend probably. Have a good weekend.

Regards,

Maarten

File: solara/components/file_drop.py [hunk 7]
```diff
@@ -33,17 +35,15 @@ def FileDrop(
     on_total_progress: Optional[Callable[[float], None]] = None,
     on_file: Optional[Callable[[FileInfo], None]] = None,
     lazy: bool = True,
-):
+) -> ipyvuetify.extra.FileInput:
```
Comment (maartenbreddels) [inline solara/components/file_drop.py:38]:
```suggestion
)
```
This is not the case, the function does not return a widget, but the type inference will (or should) figure this out, so no need to add

File: solara/components/file_drop.py [hunk 8]
```diff
@@ -94,3 +94,59 @@ def handle_file(cancel: threading.Event):
...
+def FileDropMultiple(
+    label="Drop files here",
+    on_total_progress: Optional[Callable[[float], None]] = None,
+    on_file: Optional[Callable[[List[FileInfo]], None]] = None,
+    lazy: bool = True,
+) -> ipyvuetify.extra.FileInput:
+    """Region a user can drop multiple files into for file uploading.
+
+    Almost identical to `FileDrop` except that `on_file` is called
+    with a list of `FileInfo` objects.
```
Comment (maartenbreddels) [inline solara/components/file_drop.py:109]:
```suggestion
    Almost identical to `FileDrop` except that multiple files can be dropped and `on_file` is called
    with a list of `FileInfo` objects.
```

File: solara/components/file_drop.py [hunk 9]
```diff
@@ -94,3 +94,59 @@ def handle_file(cancel: threading.Event):
...
+                    wired_files[i]["data"] = wired_files[i]["file_obj"].read()
+                else:
+                    wired_files[i]["data"] = None
+            on_file(wired_files)
+
+    result: solara.Result = hooks.use_thread(handle_file, [wired_files])
+    if result.error:
+        raise result.error
+
+    return file_drop
```
Comment (maartenbreddels) [inline solara/components/file_drop.py:152]: I don't think we want this file duplication. I'm sorry if I wasn't clear in my previous message.
Basically what I suggested is that your previous implementation could be used by these two implementations. So lets say we call your previous component `_FileDrop`, we simply call `return _FileDrop(...., multiple=True)` from this component.

**Review by maartenbreddels — CHANGES_REQUESTED** [review]
Awesome work.
Let me know if you have the energy to do these last changes. I know we're asking a lot from you, so we are happy to do the last mile.

**Comment by hkayabilisim** [comment]
No problem Maarten! I've added _FileDrop to prevent code repetitions. I hope this is what you have in your mind!

**Review by hkayabilisim — COMMENTED** [review]
It may not fit 100% to the code styling requirements but  I think my last commit https://github.com/widgetti/solara/pull/562/commits/e5e6f405fd6c48b4c33b2d43f7ad57ce757fefff solves most of the highlighted issues. 

One of the checks keeps failing but I couldn't understand why.

**Comment by iisakkirotko** [comment]
Hi @hkayabilisim! Just wanted to let you know that the CI failure is still unrelated, it should be fixed by https://github.com/widgetti/solara/pull/574. We'll try to get this PR in soon™️ :)

**Comment by maartenbreddels** [comment]
This branch had some flakiness that we wanted to attack (see the last commit) before merging it. Thanks for your patience!
